unit_scaling/tests/test_functional.py

```diff
@@ -282,23 +279,51 @@ def test_linear_scale_for_grad_input() -> None:
 
 
 def test_layer_norm() -> None:
-    input = unit_normal(2**8, 2**10)
-    weight = unit_normal(2**10)
+    input = randn(2**8, 2**10, requires_grad=True)
+    weight = randn(2**10, requires_grad=True)
     bias = zeros(2**10).requires_grad_()
     output = layer_norm(input, (2**10,), weight, bias)
     unit_backward(output)
 
     assert_unit_scaled(output, input.grad, weight.grad, bias.grad)
 
 
+# --- test add() ---
+
+
+def test_add() -> None:
+    left = randn(2**8, 2**10, requires_grad=True)
+    right = randn(2**8, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
+
+
+def test_add_broadcast() -> None:
+    left = randn(2**10, requires_grad=True)
+    right = randn(2**8, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
+
+    left = randn(2**8, 1, 2**10, requires_grad=True)
+    right = randn(2**9, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
+
+
 # --- test residual() ---
 
 
 def test_residual() -> None:
     for tau in [0.2, 0.5, 0.8]:
-        input = unit_normal(2**10)
+        input = randn(2**10, requires_grad=True)
         residual, skip = residual_split(input, tau)
-        residual = linear(residual, unit_normal(2**10, 2**10), bias=None)
+        residual = linear(residual, randn(2**10, 2**10), bias=None)
```

Comment:
No `requires_grad=True`; does that mean that this grad is not tested?

Reply:
Yeah, this linear is just to simulate having "some kind of op" on the residual branch, so checking the grads of this op's weights is probably out of scope for this test?

Reply:
Yeah, I guess I was just wondering if it helped demonstrate that the branch gradients themselves are unit scale (vs immediately scaling the backward grads at the beginning of the backward pass.) But no huge need for this.